File: 36-valid-sudoku/valid-sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function (board) {
  // checking for duplicates in rows
  for (let i = 0; i < board.length; i++) {
    const seen = new Set();
    for (let j = 0; j < 9; j++) {
      if (parseInt(board[i][j]) && seen.has(board[i][j])) {
        return false;
      } else {
        seen.add(board[i][j]);
      }
    }
  }

  // checking for duplicates in columns
  const columns = new Map();
  for (let i = 0; i < board.length; i++) {
    columns.set(
      i,
      board.map((subBoard) => subBoard[i])
    );
  }

  for (let i = 0; i < 9; i++) {
    const column = columns.get(i);
    const seen = new Set();
    for (let j = 0; j < 9; j++) {
      if (parseInt(column[j]) && seen.has(column[j])) {
        return false;
      } else {
        seen.add(column[j]);
      }
    }
  }

  //testing individual grids
  const square = new Map();
  for (let i = 0; i < 3; i++) {
    for (let j = 0; j < 3; j++) {
      square.set([i, j].toString(), []);
    }
  }
  for (let i = 0; i < 9; i++) {
    for (let j = 0; j < 9; j++) {
      const key = [Math.floor(i / 3), Math.floor(j / 3)].toString();

      square.get(key).push(board[i][j]);
    }
  }
  for (let i = 0; i < 3; i++) {
    for (let j = 0; j < 3; j++) {
      const grid = square.get([i, j].toString());
      const seen = new Set();
      for (let k = 0; k < 9; k++) {
        if (parseInt(grid[k]) && seen.has(grid[k])) {
          return false;
        } else {
          seen.add(grid[k]);
        }
      }
    }
  }

  return true;
};

File: 36-valid-sudoku/valid-sudoku.java
class Solution {
  public boolean isValidSudoku(char[][] board) {
    Set<Character> seen = new HashSet<Character>();
    for (int i = 0; i < board.length; i++) {
      for (int j = 0; j < board[i].length; j++) {
        // checking rows

        if (Character.isDigit(board[i][j]) && !seen.contains(board[i][j])) {
          seen.add(board[i][j]);
        } else if (Character.isDigit(board[i][j]) && seen.contains(board[i][j])) {
          return false;
        }
      }
      seen.clear();
    }

    // checking columns
    for (int i = 0; i < board.length; i++) {
      for (int j = 0; j < board[i].length; j++) {
        if (Character.isDigit(board[j][i]) && !seen.contains(board[j][i])) {
          seen.add(board[j][i]);
        } else if (Character.isDigit(board[j][i]) && seen.contains(board[j][i])) {
          return false;
        }
      }
      seen.clear();
    }

    // checking grid
    // create hashmap for grid
    HashMap<String, List<Character>> grid = new HashMap<String, List<Character>>();
    for (int i = 0; i < 3; i++) {
      for (int j = 0; j < 3; j++) {
        int[] gridKeyArr = new int[] { i, j };
        String gridKey = Arrays.toString(gridKeyArr);
        List<Character> gridValues = new ArrayList<>();
        grid.put(gridKey, gridValues);
      }
    }

    // populate and check grid
    for (int i = 0; i < 9; i++) {
      for (int j = 0; j < 9; j++) {
        int[] gridKeyArr = new int[] { i / 3, j / 3 };
        String gridKey = Arrays.toString(gridKeyArr);
        if (Character.isDigit(board[i][j]) && !(grid.get(gridKey).contains(board[i][j]))) {
          grid.get(gridKey).add(board[i][j]);
        } else if (Character.isDigit(board[i][j]) && (grid.get(gridKey).contains(board[i][j]))) {
          return false;
        }
      }
    }
    return true;
  }
}